File: Assignment2/Weighted_Majority.m
clear;

d = 10;
T = 1E+04;
delta = 0.1;
eta = sqrt(2*log(d)/T);

weight = ones(1, d);
agg_cost = zeros(1, d);
pseudo_regret = 0;

figure(1);

for t=1:T
    Z = sum(weight);
    weight = weight/Z;
    
    cdf = cumsum(weight);
    
    X = rand;
    I = 1;
    
    for j=1:d
        if(X <= cdf(j))
            I = j;
            break;
        end
    end
    
    cost = zeros(1, d);
    
    for j=1:d
        if(j <= d-2)
            cost(j) = binornd(1, 0.5);
        end
        if(j == d-1)
            cost(j) = binornd(1, 0.5 - delta);
        end
        if(j == d)
            if(t <= T/2)
                cost(j) = binornd(1, 0.5 + delta);
            end
            if(t > T/2)
                cost(j) = binornd(1, 0.5 - 2*delta);
            end
        end
    end
    
    Incurred_cost = weight*cost';
    
    for j=1:d
        weight(j) = weight(j)*exp(-eta*cost(j));
    end
    
    agg_cost = agg_cost + cost;
    
    pseudo_regret = max(agg_cost(I) - min(agg_cost), pseudo_regret);
    
    hold on;
    plot(t, pseudo_regret, 'r*');
end

File: Assignment2/weighted_majority.py
import numpy as np
import math

d=10
T = 100
delta = 0.1
c = np.array([0.1,0.3,0.5,0.7,0.9,1.1,1.3,1.5,1.7,1.9,2.1])
eta = c*(math.sqrt(2*(np.log10(d))/T))
regret = np.zeros(c.size)
v = np.zeros(10)

for runs in range(0,50):

	for eta_count in range(0,eta.size):

		w = np.array([1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0])
		for i in range(0,T):
			zt = np.sum(w)
			w = w/zt
			for j in range(0,8):
				v[j] = np.random.binomial(1,0.5)

			v[8] = np.random.binomial(1,0.5 - delta)
			if(i<(T/2)):
				v[9] = np.random.binomial(1,0.5 + delta)
			else:
				v[9] = np.random.binomial(1,0.5 - (2*delta))
		
			regret[eta_count] += np.dot(w,v) - np.amin(v)
			for k in range(0,10):
				w[k] = w[k] * (np.exp(-eta[eta_count]*v[k]))


	regret = regret/50

print "The regret values are",regret





